browser-extension/mv3/src/service-worker/services/requestProcessor.ts

```diff
@@ -0,0 +1,133 @@
+import { TAB_SERVICE_DATA, tabService } from "./tabService";
+
+interface AJAXRequestDetails {
+  url: string;
+  method: string;
+  type: "xmlhttprequest" | "fetch";
+  initiatorDomain: string;
+  requestHeaders: Record<string, string>;
+}
+
+class RequestProcessor {
+  private cachedRules: Record<string, Rule[]> = {
+    redirectRules: [],
+    replaceRules: [],
+  };
+
+  private updateCachedRules = async () => {
+    this.cachedRules.redirectRules = await getEnabledRules(RuleType.REDIRECT);
+    this.cachedRules.replaceRules = await getEnabledRules(RuleType.REPLACE);
+  };
+
+  constructor() {
+    this.updateCachedRules();
+    onRuleOrGroupChange(this.updateCachedRules.bind(this));
+  }
+
+  private forwardIgnoredHeadersOnRedirect = async (tabId: number, requestDetails: AJAXRequestDetails) => {
+    if (
+      !IGNORED_HEADERS_ON_REDIRECT.some(
+        (header) => requestDetails.requestHeaders[header] || requestDetails.requestHeaders[header.toLowerCase()]
+      )
+    ) {
+      return;
+    }
+
+    const matchedRule = this.findMatchingRule(
+      [...this.cachedRules.redirectRules, ...this.cachedRules.replaceRules],
+      requestDetails.url
+    );
+
+    if (!matchedRule) {
+      return;
+    }
+
+    const ignoredHeaderValues = IGNORED_HEADERS_ON_REDIRECT.map((header) => ({
+      name: header,
+      value: requestDetails.requestHeaders[header] || requestDetails.requestHeaders[header.toLowerCase()],
+    }));
+
+    const redirectedUrl = this.getRedirectedUrl(requestDetails.url, matchedRule);
+
+    return this.updateRequestSpecificRules(tabId, requestDetails.url, {
+      action: {
+        requestHeaders: ignoredHeaderValues.map((header: { name: string; value: string }) => ({
+          header: header.name,
+          value: header.value,
+          operation: chrome.declarativeNetRequest.HeaderOperation.SET,
+        })),
+        type: chrome.declarativeNetRequest.RuleActionType.MODIFY_HEADERS,
+      },
+      condition: {
+        // Exact URL is not used because for replace rules, a marker query param is add which changes the URL so exact URL doesn't match
+        urlFilter: `|${redirectedUrl}`,
+        resourceTypes: [chrome.declarativeNetRequest.ResourceType.XMLHTTPREQUEST],
+        tabIds: [tabId],
+        requestMethods: [requestDetails.method.toLowerCase() as chrome.declarativeNetRequest.RequestMethod],
+        excludedInitiatorDomains: ["requestly.io", "requestly.com"],
+      },
+    });
+  };
+
+  private findMatchingRule = (rules: Rule[], url: string): Rule | undefined => {
+    return rules.find((rule) => rule.pairs.some((pair) => matchSourceUrl(pair.source, url)));
+  };
+
+  private getRedirectedUrl = (requestUrl: string, matchedRule: Rule) => {
+    const pair = matchedRule.pairs.find((pair) => matchSourceUrl(pair.source, requestUrl));
+    const redirectedUrl = pair.destination || requestUrl.replace(pair.from, pair.to);
+    if (redirectedUrl === requestUrl) {
+      return null;
+    }
```

Comment:
Will this work in case of wildcard and regex subsitutions?

Reply:
Substitutions are not supported in case of replace rules. We need to handle only redirect rule